Stability improvements for the Marketplace tests

davehunt opened the pull request.

davehunt commented:
Hey @krupa if you get a chance to look through these that would be great. There have been a few intermittent failures in Marketplace recently, and I came up with these improvements while working on getting the tests stable enough to run on the fireplace Travis CI job. Rather than looking at the full diff I recommend looking at the individual commits. Here's an adhoc run:

http://webqa-ci-staging1.qa.scl3.mozilla.com:8080/job/marketplace.adhoc/33

davehunt commented:
Pinging @m8ttyB for review. See comment above.

bobsilverberg commented on pages/desktop/consumer_pages/base.py:
```diff
@@ -84,6 +86,29 @@ class HeaderRegion(Page):
+            return self.selenium.find_element(*self._categories_toggle_locator).text
+
+        def open_categories_menu(self):
+            categories = self.selenium.find_element(By.ID, 'header--categories')
```
This is unusual to put the locator info here, as opposed to a locator variable. Is that a leftover from some experimenting?

davehunt replied:
I think this was conscious, but I can't recall my reasons, and it breaks our convention so I'll fix it.

bobsilverberg commented on pages/desktop/consumer_pages/base.py:
```diff
@@ -84,6 +86,29 @@ class HeaderRegion(Page):
+            self.selenium.find_element(*self._categories_toggle_locator).click()
+            WebDriverWait(self.selenium, self.timeout).until(self.element_settled(categories))
+
+        class element_settled(object):
```
I like this method - it's something we have needed for some time. I wonder, though, whether it would be more appropriate to put it in `page.py`?

davehunt replied:
It would be most appropriate to put it in an `expected.py`. If you like it, I can create that now and a next step might be to extract it out to a shared library so other projects can use it. At least we can start a convention now.

bobsilverberg replied:
SGTM

bobsilverberg commented on pages/desktop/consumer_pages/base.py:
```diff
@@ -84,6 +86,29 @@ class HeaderRegion(Page):
+
+        @property
+        def categories(self):
+            return [self.Category(self.testsetup, element) for element in self.selenium.find_elements(*self._categories_locator)]
```
This line is awfully long, don't you think? ;)

davehunt replied:
I find lines like this difficult to break in a way that I am satisfied, but I'll do something about it.

bobsilverberg commented:
This LGTM @davehunt, except for a few comments which I have added. Nice work! r+ wc

davehunt commented:
@bobsilverberg I've addressed your comments. Would you mind reviewing again due to the expected.py changes. I renamed 'settled' to 'parked' as it felt like a more suitable name.

davehunt commented:
New adhoc: http://webqa-ci-staging1.qa.scl3.mozilla.com:8080/job/marketplace.adhoc/39/

davehunt commented:
I've just committed another improvement to the expected condition for moving/not moving.

bobsilverberg commented:
This looks good, @davehunt. I agree that calling it `element_not_moving` is somewhat inaccurate with the current code, as it would keep reporting false if the element had stopped moving but was still changing in size. I'm not sure how accurate and specific we need to be though.

I also wonder, though, whether these `moving` methods could just use location and not size? Are there times when an element's location is stable but its size is still changing?

Probably safest to stick with what you have here. Maybe consider changing the method names from `moving` to `changing`?

Other than that issue it all LGTM. r+

davehunt commented:
> This looks good, @davehunt. I agree that calling it element_not_moving is somewhat inaccurate with the current code, as it would keep reporting false if the element had stopped moving but was still changing in size. I'm not sure how accurate and specific we need to be though.

> I also wonder, though, whether these moving methods could just use location and not size? Are there times when an element's location is stable but its size is still changing?

If an element was hidden because it had a height of 0 and then came into view by the element heigh being increased, it would appear to be moving. Manipulating the position and size of the element are both ways to 'move' an element into view, and there may very well be more.

> Probably safest to stick with what you have here. Maybe consider changing the method names from moving to changing?

'changing' is too vague in my opinion, as this could apply to the element's style, position, size, whether it's enabled, etc. We could either be very specific `element_location_not_changing`, etc. or keep `element_not_moving` and cover additional cases as we encounter them...

davehunt commented:
Merged
